### PR TITLE
Add func to handle sparse testing data

### PR DESCRIPTION
**Pull request by WWLoveBasketball**
temp version, updating

### CODE REVIEW

File: include/LightGBM/tree.h [hunk 1]
```diff
@@ -484,6 +507,25 @@ inline int Tree::GetLeaf(const double* feature_values) const {
...
+  if (num_cat_ > 0) {
+    while (node >= 0) {
+      node = Decision(feature_values.count(split_feature_[node]) > 0 ? feature_values.at(split_feature_[node]) : 0.0f, node);
+    }
+  } else {
+    while (node >= 0) {
+      if(feature_values.count(split_feature_[node]) > 0) {
+        node = NumericalDecision(feature_values.at(split_feature_[node]), node);
+      } else {
+        node = threshold_[node] >= 0 ? left_child_[node] : right_child_[node];
```
Comment (guolinke) [inline include/LightGBM/tree.h]: I think it is better to use ```node = NumericalDecision(feature_values.count(split_feature_[node]) > 0 ? feature_values.at(split_feature_[node]) : 0.0f, node);``` .

File: src/application/predictor.hpp [hunk 2]
```diff
@@ -234,6 +265,7 @@ class Predictor {
   int num_pred_one_row_;
   int num_threads_;
   std::vector<std::vector<double>> predict_buf_;
+  std::vector<std::unordered_map<int, double>> predict_buf_map_;
```
Comment (guolinke) [inline src/application/predictor.hpp]: Not allocate for ```predict_buf_map_``` ?

File: src/application/predictor.hpp [hunk 3]
```diff
@@ -1,6 +1,9 @@
 #ifndef LIGHTGBM_PREDICTOR_HPP_
 #define LIGHTGBM_PREDICTOR_HPP_
 
+#define MAX_FEATURE 100
```
Comment (guolinke) [inline src/application/predictor.hpp]: max feature could be larger. e.g. 10k

File: src/application/predictor.hpp (hunk 2 above)
Comment (WWLoveBasketball) [reply src/application/predictor.hpp]: I have written the func 'CopyToPredictMap' to assign the map,
and I think we do not need to allocate the map.

**Comment by guolinke** [comment]
@WWLoveBasketball  you should fix this test: https://travis-ci.org/Microsoft/LightGBM/jobs/300027913

File: src/io/tree.cpp [hunk 4]
```diff
@@ -411,6 +442,40 @@ std::string Tree::NodeToIfElse(int index, bool is_predict_leaf_index) const {
   return str_buf.str();
 }
 
+std::string Tree::NodeToIfElseByMap(int index, bool is_predict_leaf_index) const {
+	std::stringstream str_buf;
+	str_buf << std::setprecision(std::numeric_limits<double>::digits10 + 2);
+	if (index >= 0) {
+		// non-leaf
```
Comment (guolinke) [inline src/io/tree.cpp]: please use 2 spaces for indent, not the tab.

**Comment by guolinke** [comment]
@wxchan  can you help to review the if-else part ?

**Pull request merged**